stdlib/nodejs/dir.rb

```diff
@@ -28,6 +30,16 @@ def entries(dirname)
       }
     end
 
+    def glob(pattern)
+      pattern = [pattern] unless pattern.respond_to? :each
+      result = Set.new
+      pattern.each do |value|
```

Comment:
@iliabylich I'm trying to write idiomatic Ruby but I don't know how to simplify this part. Maybe I should use `map`, `flatten` and `uniq` ?

```rb
pattern.map {|value| `__glob__.sync(#{value})`}.flatten.uniq
```

Reply:
What is the nesting level of `__glob__.sync(pattern)`? If it's a plain array you can use `flat_map` (which does `.map.flatten(1)`):
``` ruby
pattern.flat_map { |value| `__glob__.sync(#{value})` }.uniq
```

Also it seems that MRI doesn't call `.uniq`:
``` ruby
2.4.0 :001 > Dir.glob(['test/*', 'test/*'])
 => ["test/nodejs", "test/opal", "test/cruby", "test/nodejs", "test/opal", "test/cruby"]
2.4.0 :002 > Dir.glob(['test', 'test'])
 => ["test", "test"]
```
The interpolation is not required.
So I guess it's just a
``` ruby
pattern.flat_map { |value| `__glob__.sync(value)` }
```

Reply:
Perfect, then I will use `flat_map`. Thanks for the quick reply 👍

Reply:
> The interpolation is not required.

The String interpolation is not required ? 
As far as I know, the following code is mandatory:
```ruby
value = value.to_s if value.respond_to? :to_s
```

Otherwise an exception is thrown when `Dir.glob` is used with a `Pathname`:

```ruby
Dir.glob(Pathname.new('/foo/bar'))
```
```
TestNodejsDir#test_dir_glob:
Exception: glob pattern string required
    TypeError: glob pattern string required
        at new Minimatch (/home/dev/opal/stdlib/nodejs/node_modules/glob/node_modules/minimatch/minimatch.js:108:11)
```

Reply:
`Dir.glob` calls `to_str`, string interpolation calls `to_s`:
```
$ cat test.rb
o = Object.new
def o.to_str; p 'to_str'; end
def o.to_s;   p 'to_s';   end

Dir.glob(o)
"#{o}"

$ ruby test.rb
"to_str"
"to_s"
```

`value` variable is available in both Ruby and JS code, so you can use it as is without `#{}` in backticks (that's what I meant).

Of course, you still need to call `Opal.coerce_to!(value, String, :to_str)` to match MRI behavior, so the final should like the following (NOT TESTED):
``` ruby
def glob(pattern)
  pattern = [pattern] unless pattern.respond_to?(:each)
  pattern.flat_map do |subpattern|
    subpattern = Opal.coerce_to!(subpattern, String, :to_str)
    `__glob__.sync(subpattern)`
  end
end
```

But even this implementation is not correct, for example, for `Pathname` argument it calls `to_path` method, so I think `Opal.coerce_to!(subpattern, String, :to_s)` is better for a minimal implementation.

Reply:
Alright, I was confused because I didn't know that the `value` variable was available in both Ruby and JS code.

> But even this implementation is not correct, for example, for Pathname argument it calls to_path method, so I think Opal.coerce_to!(subpattern, String, :to_s) is better for a minimal implementation.

I could add a condition to use `to_path` if the method is available ?

```ruby
subpattern = subpattern.to_path if subpattern.respond_to? :to_path
```

Reply:
Yes, from what I see it matches the logic of MRI:
``` ruby
subpattern = subpatter.to_path if subpattern.respond_to?(:to_path)
subpattern = Opal.coerce_to!(subpattern, String, :to_str)
`__glob__.sync(subpattern)`
```

``` ruby
$o1 = BasicObject.new
$o2 = BasicObject.new

def $o1.respond_to?(*args)
  Kernel.p "$o1.respond_to?(#{args})"
  true
end

def $o1.to_path
  Kernel.p "$o1.to_path => $o2"
  $o2
end

def $o1.to_s
  '/var' # ignored
end

def $o2.respond_to?(*args)
  Kernel.p "$o2.respond_to?(#{args})"
  true
end

def $o2.to_str
  '/tmp' # return value
end

p Dir.glob([$o1])
```

Hope it helps.